nrfcache/nrfcache.go

```diff
@@ -0,0 +1,369 @@
+func (c *NrfCache) purge() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	close(c.done)
+	c.priorityQ = newNfProfilePriorityQ()
+	c.cache = make(map[string]*NfProfileItem)
+	c.evictionTicker.Stop()
+}
+
+func (c *NrfCache) startExpiryProcessing() {
+	for {
+		select {
+		case <-c.evictionTicker.C:
+			c.mutex.Lock()
+			if c.priorityQ.Len() == 0 {
+				c.mutex.Unlock()
+				continue
+			}
+
+			c.cleanupExpiredItems()
+			c.mutex.Unlock()
+
+		case <-c.done:
+			return
+		}
+	}
+}
+
+func NewNrfCache(duration time.Duration, dbqueryCb NrfDiscoveryQueryCb) *NrfCache {
+	cache := &NrfCache{
+		cache:               make(map[string]*NfProfileItem),
+		priorityQ:           newNfProfilePriorityQ(),
+		evictionInterval:    defaultCacheTTl,
+		nrfDiscoveryQueryCb: dbqueryCb,
+		done:                make(chan struct{}),
+	}
+
+	cache.evictionTicker = time.NewTicker(duration)
+
+	go cache.startExpiryProcessing()
+
+	return cache
+}
+
+type NrfMasterCache struct {
+	nrfDiscoveryQueryCb NrfDiscoveryQueryCb
+	nfTypeToCacheMap    map[models.NfType]*NrfCache
+	evictionInterval    time.Duration
+	mutex               sync.Mutex
+}
+
+func (c *NrfMasterCache) GetNrfCacheInstance(targetNfType models.NfType) *NrfCache {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	cache, exists := c.nfTypeToCacheMap[targetNfType]
+	if exists == false {
+		fmt.Printf("Creating cache for nftype %v", targetNfType)
+		cache = NewNrfCache(c.evictionInterval, c.nrfDiscoveryQueryCb)
+		c.nfTypeToCacheMap[targetNfType] = cache
+	}
+	return cache
+}
+
+func (c *NrfMasterCache) clearNrfMasterCache() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	for k, cache := range c.nfTypeToCacheMap {
+		cache.purge()
+		delete(c.nfTypeToCacheMap, k)
+	}
+}
+
+func (c *NrfMasterCache) removeNfProfile(nfInstanceId string) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	var ok bool
+	for _, nrfCache := range c.nfTypeToCacheMap {
+		if ok = nrfCache.removeByNfInstanceId(nfInstanceId); ok {
+			break
+		}
+	}
+	return ok
+}
+
+var masterCache *NrfMasterCache
+
+type NrfDiscoveryQueryCb func(nrfUri string, targetNfType, requestNfType models.NfType, param *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (models.SearchResult, error)
+
+func InitNrfCaching(interval time.Duration, cb NrfDiscoveryQueryCb) {
+	m := &NrfMasterCache{
+		nfTypeToCacheMap:    make(map[models.NfType]*NrfCache),
+		evictionInterval:    interval,
+		nrfDiscoveryQueryCb: cb,
+	}
+	masterCache = m
+}
+
+func disableNrfCaching() {
+	masterCache.clearNrfMasterCache()
+	masterCache = nil
+}
+
+func SearchNFInstances(nrfUri string, targetNfType, requestNfType models.NfType, param *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (models.SearchResult, error) {
+
+	var searchResult models.SearchResult
+	var err error
+
+	c := masterCache.GetNrfCacheInstance(targetNfType)
+	if c != nil {
+		searchResult, err = c.handleLookup(nrfUri, targetNfType, requestNfType, param)
+	} else {
+		fmt.Println("SearchNFInstances nrf cache")
+
```

Comment:
Why did the message changed. In the `nrf`, the message says "Failed to find cache for nftype". Should not the message here be kind of similar? or am I overlooking somthing?
```suggestion
		fmt.Println("failed to find cache for nftype")
```

Reply:
This is my mistake. The original log message is used.